### ISSUE TITLE
Adjusting autosubmit parameters for the Genetic Algorithm (GA) parameter optimization workflow in AS 3.14.0

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#979]
In GitLab by @mgali on Mar 17, 2023, 11:53

In this issue I plan to document all the adjustments needed in the Autosubmit workflow (specifically wrappers, processors per core, etc.) for optimal performance in the Genetic Algorithm (GA) parameter optimization experiments.

The workflow and typical experiments are described in [this paper by Marcus Falls et al. 2022](https://gmd.copernicus.org/articles/15/5713/2022/gmd-15-5713-2022.html).

The autosubmit templates are in [this repository](https://earth.bsc.es/gitlab/cp/genetic_algorithm_pisces1d/-/tree/master).

In a typical GA experiment, an initial simulation (INI, template `pre.sh`) is followed by a generation where 100 simulations (SIM, template `sim.sh`) with different parameter sets run in parallel. Each of these simulations (individuals) is a CHUNK and runs on 1 core in less than 5 minutes. The outputs of each simulation are collected by the POST job (template `post.sh`) to compute a cost function (based on different goodness-of-fit statistics) and select the refined parameter sets that will be used in the subsequent 100 generations. This process is repeated iteratively for N generations (= MEMBERS, typically no less than 40 and a maximum of 100) to obtain the optimal parameter set.

In this GA workflow developed by Marcus Falls, MEMBERS (= generations) are vertical and CHUNKS (= individual simulations with a given parameter set) are horizontal. In [this issue](https://earth.bsc.es/gitlab/cp/genetic_algorithm_pisces1d/-/issues/9) Marcus documented his tests with horizontal-vertical wrappers in AS 3.13.0, which were not successful. Horizontal wrappers only did work in AS 3.13.0.

Now we need to run more experiments but adjustments to the wrappers and other AS parameters seem necessary, as we found when running a similar workflow used for sensitivity analyses documented [in this GA issue](https://earth.bsc.es/gitlab/cp/genetic_algorithm_pisces1d/-/issues/10) and [in this autosubmit issue](https://earth.bsc.es/gitlab/es/autosubmit/-/issues/941).

I re-launched experiments a3tg (horizontal-vertical wrappers) and a41f (horizontal only) as a starting point for debugging. Both experiments had run smoothly in AS 3.13.0 in November 2021, although the addition of vertical in a3tg made no difference in practice.

Tagging @mcastril and @dbeltrankyl , who already did debugging work for the sensitivity analyses. Keeping in the loop @ rbernard , @ macosta and @jllort. Thanks to all!

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 20, 2023, 07:43

Update: after launching **a3tg** and **a41f** on Wednesday 15/03 afternoon, I see that in both experiments INI (= the first SIM with default parameters) completed on 18/03 with no apparent errors after queueing for over 2 days. However, in Autosubmit the INI job still shows as queueing and the experiments went inactive. Any idea why this happened?

I re-launched a41f aftter marking INI as completed to see how the horizontal-wrappers-only workflow proceeds.

I left a3tg untouched.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 20, 2023, 08:19

Hello @mgali ,

There is a bug in the wallclock function (fixed on 3.15.0b.

Autosubmit looks to the wallclock timer of every running job and puts them in to failed status if they are over the wallclock specified even if there is no scheduler.

This is correct and to solve it, you just have to put an higher wallclock to the ini job the bug is that the job is not cancelled in the remote.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 20, 2023, 10:40

Thanks @dbeltrankyl , will take care ofm aking that change in the following experiments

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 20, 2023, 11:20

Now that experiments are proceeding, let's see how they work:

- a3tg: with current horizontal-vertical wrappers (inherited from Marcus' Falls work 1.5 years ago) the experiment essentially runs as sequential: there's only 1 SIM running at a time, with 1 submitted and several queueing (the queueing list seems to grow while the 100-CHUNKS MEMBER is advancing).

- a41f: with current horizontal wrappers inherited too from the old experiment, I see 3-5 SIM running at a time, meaning it takes around 1.5 hours to complete one MEMBER (theoretically this could take less than 5 minutes, with current queueing times of less than 2 minutes).

My plan:

Launch experiment **a41g** with horizontal wrappers adding the following settings:

- autosubmit*.conf: `MAX_WRAPPED_H = 48` (possibly unnecessary because maximum wrapped jobs is default behaviour?),  `MIN_WRAPPED_H = 25` (trying my luck here to see how quickly can 25 jobs can be allocated, ie 1/4 of a MEMBER)
- jobs*.conf: `TASKS = 1`

For the moment not trying vertical wrappers, nor the `CUSTOM_DIRECTIVES = ["#SBATCH --exclusive"]` in platforms (as done in the related sensitivity analysis workflow).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 21, 2023, 11:59

> autosubmit\*.conf: `MAX_WRAPPED_H = 48` (possibly unnecessary because maximum wrapped jobs is default behaviour?), `MIN_WRAPPED_H = 25` (trying my luck here to see how quickly can 25 jobs can be allocated, ie 1/4 of a MEMBER)

Yes, the maximum if not specified will be until platform MAX_WALLCLOCK is satisfied or max cores allocated ( 720 debug, 2400 bsc_queue) If I recall correctly

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 22, 2023, 08:31

Now three experiments have been running for nearly 2 days:

- a3tg has completed 3 generations with 100 sims each and seems to be running sequentially
- a41f has completed 19 generations and 2-4 sims are generally running at a time
- a41g has completed 4 generations

So the horizontal-wrappers-only configuration inherited from 2021 experiments still works best. With this setup it is possible to run one experiment in about 4 days, which is sufficient for our needs.

Increasing MIN_WRAPPED_H increases the queue time needed to allocate sims to nodes, and this is not compensated by savings in greater number of sims running simultaneously.

I am going to kill a3tg and a41g and use them to make a further test before starting to run "real" experiments.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 23, 2023, 07:43

New + ongoing tests, one day later, all on `bsc_es` queue:

- **a41f** with horizontal wrappers and Autosubmit defaults ("old" config) has completed 27 MEMBERS (generations) / 3 days; just 2-3 SIMs were running at a time the most times I checked.
- **a41g** with horizontal wrappers, `autosubmit` MIN_WRAPPED_H = 5, `jobs` [SIM] TASKS = 1, `platforms` [marenostrum4] PROCESSORS_PER_NODE = 5, has completed 16 MEMBERS / 1 day; up to 11 simultaneous SIMs, usually >5.
- **a41h** identical to a41g with addition in `platforms` [marenostrum4] CUSTOM_DIRECTIVES = ["#SBATCH --exclusive"], has completed <3 MEMBERS / 1 day; running sequentially.
- **a3tg** with horizontal wrappers as a41f, `jobs` [SIM] TASKS = 1,  `platforms` [marenostrum4]  PROCESSORS_PER_NODE = 1 and CUSTOM_DIRECTIVES = ["#SBATCH --exclusive"], has completed 3.5 MEMBERS / 1 day; generally 1-2 simultaneous SIMs, sometimes more.


So far **a41g** working best. Will further tune it in 3 additional exps where both MIN_WRAPPED_H and PROCESSORS_PER_NODE will be set to 3, 8 and 12.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 23, 2023, 10:56

New exps running after killing old instances of a3tg and a41h:

- **a41h**: as a41g but MIN_WRAPPED_H = 3 and PROCESSORS_PER_NODE = 3
- **a41i**: as a41g but MIN_WRAPPED_H = 8 and PROCESSORS_PER_NODE = 8
- **a41j**: as a41g but MIN_WRAPPED_H = 12 and PROCESSORS_PER_NODE = 12

Launched 23/03 between 9-10 am.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 23, 2023, 16:28

In a41g, member 0018 (generation 19) has 3 failed SIMs

- 0007: ` ===>>>> : bad opening file: namelist_pisces_cfg`
- 0019: empty runtime
- 0021: empty runtime

In exps submitted in 23/03 morning, a41h-i-j, there's no EXP00 folder in exp directory on SCRATCH. Weird? May need to relaunch.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 24, 2023, 07:56

The experiment with a41g-j didn't work well. To understand what happened I kept them in experiment directories named `a41?_old` on my **MN4 scratch**, and kept also logs called `log1` in **esarchive/autosubmit** directory experiments.

In the case of a41g the exp seemed to progress quickly and the SIMs were being marked as completed but around half apparently had not run (the rest had run successfully). In consequence the parameters folder had just the csv file of the first generation inside. These messages started appearing in the log (file `log1`) from the first generation (member):

```
4041 of 4041 jobs remaining (12:12)
Job a41g_INI is RUNNING


4041 of 4041 jobs remaining (12:13)
^[[32mJob a41g_INI is COMPLETED^[[0m^[[39m
Wrapper policy is set to flexible and there is a deadlock, As will submit the jobs sequentally
Call from a41g_19600101_fc0000_1_SIM with status SUBMITTED


4040 of 4041 jobs remaining (12:13)
Job a41g_19600101_fc0000_1_SIM is RUNNING
Wrapper policy is set to flexible and there is a deadlock, As will submit the jobs sequentally
Call from a41g_19600101_fc0000_2_SIM with status SUBMITTED


4040 of 4041 jobs remaining (12:14)
Job a41g_19600101_fc0000_2_SIM is QUEUING
```

In the case of a41h,i,j g some SIMs were being marked as completed but in reality the the EXP00 folder didn't even exist.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 24, 2023, 07:58

On 24/03 I decide to:

- Relaunch a41g with `POLICY = MIXED` in `autosubmit` [wrapper] (instead of the default `FLEXIBLE`). Not much hope it works. If it does I will run the experiments with variable PROCESSORS_PER_NODE and MIN_WRAPPED_H.
- Relaunch a41h,i,j with same wrappers as a41f to get the real experiment done.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 24, 2023, 09:48

Hi @dbeltrankyl , @mcastril 

I don't know what's going on with the GA workflow in autosubmit and it is very frustrating.

Experiment **a41f** had run smoothly (though a bit slowly) for 34 members (= GA generations) I just noticed that after one job failed, all the results from the previous 34 generations stored in folders `parameters`, `sPOC` and `bPOC` have been deleted. In these folders there should be 1 file per generation, which allows me to analyze the optimized parameters'  evolution.

This means I cannot even re-start the experiment from generation 34 and have to start from zero again.

More generally: 1.5 years back I was able to run 10-15 simultaneous experiments of this kind, each with their 40 MEMBERS and 100 CHUNKS in each member. Now I can't even complete one...

I will really appreaciate that you engage in this debugging work. Otherwise a lot of efforts will be wasted! Cheers

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 24, 2023, 10:06

Hello @mgali , 

I was just looking at it, 

There is a misconfiguration in your experiment ( atleast autosubmit_a41h) 

It seems like you did a manual copy from other's experiments.

Please, take in mind that if you copy manually the experiment configuration files, it will not update the EXPID from the configuration files.

```
In autosubmit_a41h.conf I see:

 
[config]
# Experiment identifier
# No need to change
EXPID = a41f

In expdef_a41h.conf I see 

dbeltran@ bscearth357:/esarchive/autosubmit/a41h/conf> cat expdef_a41h.conf 
[DEFAULT]
# Experiment identifier
# No need to change
EXPID = a41h
```

This should be a41h, because autosubmit will look in the remote for the a41f folder instead of the correct one.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 24, 2023, 10:09

* Same for a41j autosubmit_a41j.conf
* And same for a41i autosubmit_a41i.conf

All of them are pointing to a41f

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 24, 2023, 10:28

I just added a new issue https://earth.bsc.es/gitlab/es/autosubmit/-/issues/994

to solve it for 3.15.0b and 4.0 so it doesn't happen anymore

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 24, 2023, 10:46

If I understood well and if there is a bug in the automatic cancellation of tasks exceeding the workflow (only the task, not the whole job as I understand) I think this is worth a patch in 3.14.0 because it can save significant computation time instead of freezes, as we had in the past.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 24, 2023, 10:49

Thanks @dbeltrankyl ! I had checked by doing greps with expid, but I probably missed something because the ones I checked were fine. Let's see if this solves the issues...

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 24, 2023, 10:54

@mgali in #941 we determined that something could have changed in the MPI configuration that doesn't allow to run 48 MPI executions in the same node. 

You were using:

```
[marenostrum4]
PROCESSORS_PER_NODE = 24
```

And it seemed to work better. Could you try allocating the double of nodes and allocating half of their cores with the horizontal wrapper?

@dbeltrankyl do you think that you can copy one of @mgali experiments and see it everything it's ok with the horizontal-vertical?

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 24, 2023, 11:45

Thanks @mcastril . Some remarks:

- The experiment a41f that was working fine had `PROCESSORS_PER_NODE = 48` .
- In other experiments I had started testing reduced `PROCESSORS_PER_NODE` numbers, from 3 to 12. I'll resume these tests.

Of course I'll appreciate it very much if @dbeltrankyl can test an horizontal-vertical.

Thanks to both for inputs and test!

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 24, 2023, 11:49

I see what I did wrong, I didnot realise EXPID was also in autosubmit*.conf, not just in expdef*.conf.

Re-launching everyting with corrections. Thanks

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 24, 2023, 16:01

Update on 20230324: to make sure I'm not messing up with the experiments, I created 5 new ones, **a5t[qrsuv]**.

The settings of a5tq correspond to old a41f. The others correspon to the various tests with `PROCESSORS_PER_NODE` and `MIN_WRAPPED_H` with values 3, 5, 8, 12.

Let's see how things go over the weekend.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 27, 2023, 08:52

Summary of performance metrics as of 20230327 8:30:

| EXPID | [autosubmit] wrappers | [autosubmit] `MIN_WRAPPED_H` | [platforms] `PROC._PER_NODE` | [jobs] `TASKS` | gen./hour | SYPD | ASYPD | Total queue time (h) | Jobs failed |
| ------ | ------ | ------ | ------ | ------ | ------ | ------ | ------ | ------ | ------ |
| a5tq | horizontal | - | 24 | - | 0.47 | 0.22 | 0.11 | 187 | 66 |
| a5tr | horizontal | 3 | 3 | 1 | 0.97 | 0.21 | 0.11 | 496 | 0 |
| a5ts | horizontal | 5 | 5 | 1 | 1.14 | 0.21 | 0.12 | 175 | 0 |
| a5tu | horizontal | 8 | 8 | 1 | 1.14 | 0.21 | 0.12 | 167 | 0 |
| a5tv | horizontal | 12 | 12 | 1 | 1.14 | 0.21 | 0.12 | 155 | 0 |

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 27, 2023, 08:54

Note: I had to relaunch a5ts,u,v after 12h before they had gone inactive after initial launch. They ran very smoothly after that.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 28, 2023, 13:42

> @dbeltrankyl do you think that you can copy one of @mgali experiments and see it everything it's ok with the horizontal-vertical?

I'll try to check this soon, I had other issues pending

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 31, 2023, 08:14

I invite you all @mcastril @dbeltrankyl @kinow to keep an eye on this issue. **It seems some recent changes autosubmit are making it very difficult for me to run this particular workflow**.

Update: I launched 5 new experiments on 28/03 morning. They had exactly the same configurations as a5tr (new a5u9, a5ub), a5ts (new a5ua), a5tu (new a5uc) and a5tv (new a5ud). I hoped they would run as smoothly, but they did not.

All the new experiments went inactive repeatedly with this error:
```
[CRITICAL] Check job and queue specified in jobs.conf. Sections that could be affected: SIM [eCode=7014]
```

I realized this same error message had appeared in the 3 experiments that had gone inactive in the previous set: a5ts, a5tu, a5tv. HOWEVER, this error had not prevented me from relaunching and completing them successfully. Note that this error had not appeared in previous experiments run with autosubmit 3.14 earlier this month.

A further new error appeared in the newest set of experiments, which this time seems to prevent me from relaunching them:
```
There is a bug in the code, please contact via git [eCode=7070]
```

I also checked https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/311 and realized some tests in autosubmit 3.15 have been propagated to 3.14.

This may be related to a new error message that appears in my logs since the last few days. Example:
```
ERROR:paramiko.transport:Secsh channel 97 open FAILED: open failed: Administratively prohibited
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 31, 2023, 09:15

Hello,
 the update is not yet in the module. I will check what happens but it is not related to an update

https://earth.bsc.es/gitlab/es/autosubmit/-/tags/v3.14.3 is the last update (1 month )

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 31, 2023, 09:38

> HOWEVER, this error had not prevented me from relaunching and completing them successfully.

`06:56:19,423 Trace: sbatch: error: qosmaxsubmitjobperuserlimit`

Seems that you have more jobs active as allowed by Marenostrum4 policies, this is explained by the following:

I see that the relaunch causes to send multiple jobs sequentially due to the policy of the wrapper being flexible and the min_wrapped being no longer valid as flexible doesn't wait for all to be balanced again, I recommend you use mixed or strict, so this does not happen. 



> `ERROR:paramiko.transport:Secsh channel 97 open FAILED: open failed: Administratively prohibited`

This is a temporal ban from marenostrum, probably due to the amount of jobs logs being recovered at once as each of them makes a new connection, maybe you should limit the total jobs or launch different experiments with a bit of delay so the jobs don't end strictly at the same time.

Well, It can also be due `06:56:19,423 Trace: sbatch: error: qosmaxsubmitjobperuserlimit`

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 31, 2023, 09:40

> `There is a bug in the code, please contact via git [eCode=7070]`

About this, which experiment has this error? But may be related to the max jobs per user from marenostrum4

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 31, 2023, 09:40

So then there's been changes somewhere else that cause new errors and eventually stop experiments. It's all completely out of my expertise, but from my user perspective makes running these experiments quite frustrating, after having spent quite  abit of time on tyring to debug and optimize the workflow.

BTW, these messages are recorded in the log* files in /esarchive/autosubmit/$EXPID/. Some experiments have just one, some have log1, log2.

Thanks a lot for checking.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 31, 2023, 09:42

> About this, which experiment has this error? But may be related to the max jobs per user from marenostrum4

At least a5ud.

Probably in it's close relatives a5u[9,a,b,c], but need to re-check.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 31, 2023, 09:46

True, I see the jobs are no longer wrapped, making us hit the maximum jobs allowed.

I'll add `POLICY = MIXED` in [autosubmit][wrappers]. Is that fine?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 31, 2023, 09:47

Yes, The difference between mixed and strict is that mixed can retry failed jobs ( it will sent them sequentially ) and strict no

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 31, 2023, 09:50

Another remark: after a5u[9,a,b,c] had repeatedly gone inactive, and having seen this error:
```
[CRITICAL] Check job and queue specified in jobs.conf. Sections that could be affected: SIM [eCode=7014]
```

I wondered whether this could be fixed by running jobs INI and POST on bsc_es, because so far these jobs were on the debug queue. So I restarted the experiments after modifying jobs.conf. This restart option is possible in this workflow: just need to retake experiments where they stopped and specify restart=T in the `pre.sh` template. 
So if you compare the jobs.conf of, say, a5tr and a5ud, you'll see this difference.

But apparently this had nothing to do with my issues. Just keeping record of everything here.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 31, 2023, 10:02

To clarify and tl;dr. 

Autosubmit 3.14 -> stable version -> No changes unless it is a vital fix, but all are announced in the mail list. There will be an update next week ( and the reason you saw some commits in 3.14 branch ):

* Fix overclock jobs 
* Fix an issue when multiple jobs are submitted simultaneously, but one has QOS error. ( I think this only happens with different sections, but maybe you could check the squeue) What happens there is that the non-QoS failed jobs are submitted without AS notice them

Autosubmit 3.15.0 -> beta version -> Last autosubmit 3 version will include support for new platforms and some fixes but not new implementations


The main issue: qosmaxsubmitjobperuserlimit derived from an imbalanced wrapper and a big quantity of jobs being launched from the same user.

> Another remark: after a5u\[9,a,b,c\] had repeatedly gone inactive, and having seen this error:

These jobs are inactive due quosmaxsubmitjobperuserlimit is not related to AS configuration but rather to the platform limiting the number of jobs.. 


I never saw this error:

```
[ERROR] 2023-03-30 16:18:46,553 Trace: can't start new thread
[CRITICAL] 2023-03-30 16:18:46,553 There is a bug in the code, please contact via git [eCode=7070]
```
For me seems like an escalation issue or no enough resources in Autosubmit Virtual Machine 


Seems related to the mentioned, but in Autosubmit Virtual Machine, maybe you can use both machines instead of one to avoid this issue?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 31, 2023, 10:10

In other news,

I'm testing some changes to horizontal-vertical machinesfiles written using a copy of a5tv. There are at least two 
 models that have issues with these machine files

If I'm successful, the changes should be in the 3.15.0b, but we can see other solutions like updating devel module derived from 3.14.0 instead of 3.15 , in any case is still working on

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 31, 2023, 10:17

Thank you @dbeltrankyl. I think that you testing the workflow yourself is the only way to figure out what's going on with the wrappers.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 31, 2023, 10:25

@mgali thank you for your patience with these issues, I understand your frustration.

With the release of Autosubmit 4 (still beta), parallel to AS 3.14, and an increasing number of users, there's more activity than usual. Additionally, issues involving workflows with high task throughput launched concurrently are usually more difficult to debug, since there are external factors in play, like platform policy and environment restrictions. 

That's why I thought that the best way is to test the workflow ourselves, but this needs some time dedication that Dani is trying to allocate.

Thanks again

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 31, 2023, 10:33

Hi @mcastril and @dbeltrankyl ,

Thanks a lot for your inputs. I am aware that I am dealing with a workflow that is prone to complications. I don't have the knowledge needed to optimize it so the tests I make are quite empirical. 

To let you better understand the situation, there's different needs with different time horizons:

- Over the next few months I would like to complete a set of experiments I started 2 years back with Marcus Falls. It's a small number compared to those we already completed in 2021. That's why I keep running experiments in "production mode", with large numbers of parallel sims, while testing alternative autosubmit parameters to eventually make them run faster.

- On the mid-term, say 3-4 years, I am planning to use the genetic algorithm workflow in other projects, if granted. Other colleagues (eg @jllort , @ rbernard ) are also interested in running this kind of experiments in the future. A proper optimization like the one you are atempting @dbeltrankyl will be very useful in this regard.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 31, 2023, 10:53

>Yes, The difference between mixed and strict is that mixed can retry failed jobs ( it will sent them sequentially ) and strict no

I tried MIXED wrappers and got this error:

```
[CRITICAL] Wrapper couldn't be formed under mixed POLICY due minimum limit not being reached: [wrappeable:1 < defined_min:5] [wrappeable_h:1 < defined_min_h:5]|[wrappeable_v:1 < defined_min_v:1]
```

Will try STRICT.

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 31, 2023, 11:15

I tried STRICT and got a similar error:
```
[CRITICAL] Wrapper couldn't be formed under strict POLICY due minimum limit not being reached: [wrappeable:1 < defined_min:5] [wrappeable_h:1 < defined_min_h:5]|[wrappeable_v:1 < defined_min_v:1]
Check your configuration: Only jobs_in_wrappers are active, check their dependencies. [eCode=7014]
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 31, 2023, 11:16

Which experiment is it? To see the current workflow status

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 31, 2023, 11:29

Sorry, I recycled those experiments to launch them with the old horizontal wrappers config. The error message for "MIXED" can be found in `/esarchive/autosubmit/a5v4/log`.

Can make a copy and launch with MIXED or STRICT if you want to check... it seems MIN_WRAPPED_H = 5 clashes with MIXED and STRICT wrapper policies.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 31, 2023, 11:52

Hello @mgali ,

I've figured the issue, 

Could you use:

```
[wrapper]
TYPE = horizontal
JOBS_IN_WRAPPER = SIM
MACHINEFILES = STANDARD
#MIN_WRAPPED_H = 5
#MAX_WRAPPED_H = 5
MIN_WRAPPED = 5
MAX_WRAPPED = 25
POLICY = STRICT
```

The WRAPPED_H/V is bound to horizontal-vertical and vertical-horizontal, I saw that it had an erratic behavior when used with vertical/horizontal for this experiment.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 31, 2023, 11:52

well the min/max can be other values is the one I just tested

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 31, 2023, 11:54

The issue is that the MAX_H considers the SIM dependencies ( post-1 ). To calculate the max amount of jobs and for horizontal  wrapper there is no need. MAX without V/H never takes this into consideration. 

I'll add in the autosubmit checker to show a critical message if H/V is used with 1D wrappers

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 31, 2023, 11:57

By the way, this is probably what caused to don't do wrappers in the other experiments, so flexible policy should also work.

EDIT: You can always check the wrappers with inspect -cw or monitor -cw before running the experiment

Edit2: With inspect, you only have to check the TMP folder and look for ASTRHREAD cmd, the name of the cmd indicates the amount of jobs inside the wrapper ) So you don't have to wait for the plot using monitor -cw

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 3, 2023, 11:54

Update: 

* Machinefiles rewrote not enough, same issues(perhaps it went a bit better but same "job step creation issue"). But at least now all jobs will have specified their nodes and exact cores, so I think it is better anyway. 

* Erased all Threads from wrapper and instead did an equivalent using subprocess testing it.

Edit: The problem with the threads is that I would need to do the same for the other wrapper if it works. 

~~Edit2: The status checker inside the wrapper is sequential, I think it was already like that, it only affects the log recovery threads and shouldn't impact it too much unless unbalanced wallclock in the horizontal level.~~ ( NVM, the complete file is created once the job is finished, discard this edit )

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 3, 2023, 12:05

Okay, I completed 1 horizontal level + POST without issues. Now it is running the second SIM level and no job step creation issues. 

I think this solution is working, but the code is not yet inside Autosubmit as I'm "hacking" autosubmit to use a custom wrapper instead the one it should launch for now.

Running 5 members only, 5 chunks.

| MEMBER | FINISH_TIME |
| ------ | ----------- |
| 0 | 12:00 |
| 1 | 12:14 |
| 2 | 12:28 |
| 3 | 12:42 |
| 4 | 12:57 |
| 5 | 13:11 |

I notice that:

Each horizontal job starts at the same time.

But between each job horizontal, the completed time is delayed is this normal? @mgali.

Example, current member running:

```
-rw-r--r-- 1 bsc32070 bsc32   11 Apr  3 12:28 a000_19600101_fc0003_1_SIM_STAT
-rw-r--r-- 1 bsc32070 bsc32   11 Apr  3 12:28 a000_19600101_fc0003_3_SIM_STAT
-rw-r--r-- 1 bsc32070 bsc32   11 Apr  3 12:28 a000_19600101_fc0003_4_SIM_STAT
-rw-r--r-- 1 bsc32070 bsc32   11 Apr  3 12:28 a000_19600101_fc0003_2_SIM_STAT

```

```
-rw-r--r-- 1 bsc32070 bsc32    0 Apr  3 12:36 a000_19600101_fc0003_3_SIM_COMPLETED
-rw-r--r-- 1 bsc32070 bsc32    0 Apr  3 12:37 a000_19600101_fc0003_2_SIM_COMPLETED
-rw-r--r-- 1 bsc32070 bsc32    0 Apr  3 12:39 a000_19600101_fc0003_1_SIM_COMPLETED
-rw-r--r-- 1 bsc32070 bsc32   22 Apr  3 12:42 a000_19600101_fc0003_4_SIM_COMPLETED

```

Rest of logs for next member:

```
-rw-r--r-- 1 bsc32070 bsc32    0 Apr  3 12:42 a000_19600101_fc0003_POST.cmd.out.0
-rw-r--r-- 1 bsc32070 bsc32    0 Apr  3 12:42 a000_19600101_fc0003_POST_COMPLETED
-rw-r--r-- 1 bsc32070 bsc32   22 Apr  3 12:42 a000_19600101_fc0003_POST_STAT
-rw-r--r-- 1 bsc32070 bsc32  47K Apr  3 12:42 a000_19600101_fc0003_POST.cmd.err.0

-rw-r--r-- 1 bsc32070 bsc32    0 Apr  3 12:42 a000_19600101_fc0004_1_SIM.cmd.out.0
-rw-r--r-- 1 bsc32070 bsc32    0 Apr  3 12:42 a000_19600101_fc0004_4_SIM.cmd.out.0
-rw-r--r-- 1 bsc32070 bsc32    0 Apr  3 12:42 a000_19600101_fc0004_3_SIM.cmd.out.0
-rw-r--r-- 1 bsc32070 bsc32    0 Apr  3 12:42 a000_19600101_fc0004_2_SIM.cmd.out.0
-rw-r--r-- 1 bsc32070 bsc32   11 Apr  3 12:42 a000_19600101_fc0004_1_SIM_STAT
-rw-r--r-- 1 bsc32070 bsc32   11 Apr  3 12:42 a000_19600101_fc0004_3_SIM_STAT
-rw-r--r-- 1 bsc32070 bsc32   11 Apr  3 12:42 a000_19600101_fc0004_4_SIM_STAT
-rw-r--r-- 1 bsc32070 bsc32   11 Apr  3 12:42 a000_19600101_fc0004_2_SIM_STAT
-rw-r--r-- 1 bsc32070 bsc32  35K Apr  3 12:43 a000_19600101_fc0004_1_SIM.cmd.err.0
-rw-r--r-- 1 bsc32070 bsc32  35K Apr  3 12:43 a000_19600101_fc0004_4_SIM.cmd.err.0
-rw-r--r-- 1 bsc32070 bsc32  35K Apr  3 12:43 a000_19600101_fc0004_2_SIM.cmd.err.0
-rw-r--r-- 1 bsc32070 bsc32  35K Apr  3 12:43 a000_19600101_fc0004_3_SIM.cmd.err.0

```

Is it just to see if it is a downside of the subprocess or is the code itself

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 3, 2023, 12:07

I need to check the cores value after this test is finished, right now I put 8, but I'm launching only 4 sims. I will check if 4 + 1 is enough. ( 1 for the main job, 4 for PROCESSORS*TASKS= 1 ).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 3, 2023, 12:14

About the administrative bans and horizontal wrappers.. maybe we can implement 3.14-3.15) and sleep(chunk_number) or sleep(random(0,chunk_level)) before the thread connects to the platform? This is a bad fix but may help for Autosubmit 3 @mcastril.

This will avoid making 100 connections simultaneously if the jobs inside the wrappers are 100 at the horizontal level, per example.

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 3, 2023, 19:31

> I see that the relaunch causes to send multiple jobs sequentially due to the policy of the wrapper being flexible and the min_wrapped being no longer valid as flexible doesn't wait for all to be balanced again, I recommend you use mixed or strict, so this does not happen.

Dani, but I understand that `flexible` is not the best solution in this scenario, because it's harder to control de minimum number of jobs to wrap, right?

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 3, 2023, 19:38

> Erased all Threads from wrapper and instead did an equivalent using subprocess testing it.

I didn't understand it @dbeltrankyl 

Good to see that you are progressing!

> About the administrative bans and horizontal wrappers.. maybe we can implement 3.14-3.15) and sleep(chunk_number) or sleep(random(0,chunk_level)) before the thread connects to the platform? This is a bad fix but may help for Autosubmit 3 @mcastril.

This is for the log recovery, right? 

Some ideas, without knowing the internals as you do:

- If N jobs are completed in the same cycle (safety check...), retrieve M of them in N/M batches, leaving some time in between. How can you do this if the safety time is small? Do you have to introduce a second thread to in turn span different threads?
- The other option is that Autosubmit leaves N-M logs for the next cycle, but I don't know if that can be done technically

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 4, 2023, 08:49

Right, I think that the best would be mixed

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 4, 2023, 09:03

> I didn't understand it @dbeltrankyl 

Right now:

The wrappers use threads to launch the jobs.
These threads count as used cores that are obtained from the wrapper allocation. So for launch jobs you would need to allocate a core per thread and then another one for the mpirun.

The amount of cores used for 5 jobs is:
1 wrapper job + 1 joblist thread + 5 threads + 5 mpirun

The new code:

Reduces this to 5 mpiruns cores ( or 6 one for the wrapper job, not sure have to test it )

Subprocesses use the wrapper job ( main job ) resources, so I changed these threads per the subprocess lib, and now each mpirun has its appropriate resources by looking at the machines files.
I'm using the process.wait() equivalent to thread.start() thread.join() to run the apropiated jobs.

My only concern is if this is affecting the job performance as indicated in one of the comments ( the table with the ending dates )

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 4, 2023, 09:55

> Some ideas, without knowing the internals as you do:

The problem right now:

* Threads are generated at job level each time a job reaches the failed or completed status. 
* Threads only known info about their associated job. 

I wanted to do a non-invasive change as we need to refactor it for AS 4 anyway, so: 

> * If N jobs are completed in the same cycle (safety check...), retrieve M of them in N/M batches, leaving some time in between. How can you do this if the safety time is small? Do you have to introduce a second thread to in turn span different threads?

I can't put a delay in the main job, as it will affect the performance, and once generated, they know nothing about the other threads so not sure how to do the synch without doing major changes that will be unnecessary for AS 4. 

To do this, we would need to change the thread from job level to job_list or Autosubmit level that communicates with the Main processor each time any job reaches the FAILED or COMPLETED status and basically, this is the first procedure to do in AS 4 refactor 

* The other option is that Autosubmit leaves N-M logs for the next cycle, but I don't know if that can be done technically

This is not possible right now, as the threads are spawned at the job level. 

My simple idea for AS 3 was to rely on data that is easily accessible from the thread without needing to do a refactor in AS 3.

edit: And yes this is for the log recovery, stats, and jobid procedures

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 5, 2023, 18:22

> The wrappers use threads to launch the jobs. These threads count as used cores that are obtained from the wrapper allocation. So for launch jobs you would need to allocate a core per thread and then another one for the mpirun.

> The amount of cores used for 5 jobs is: 1 wrapper job + 1 joblist thread + 5 threads + 5 mpirun

Thank you @dbeltrankyl. It's strange that we didn't notice until now. Something must have changed in the way that the threads allocate resources... MONARCH was working perfectly with horizontal and horizontal-vertical wrappers.

> My only concern is if this is affecting the job performance as indicated in one of the comments ( the table with the ending dates )

Good that you found a solution. I suppose that we don't expect the simulations to be so different in duration. Would it mean that there is a serialization in the main process?

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 6, 2023, 10:12

Do you know why Python threads are consuming additional cores @dbeltrankyl ?

Were you able to reproduce the behavior locally? Maybe it's the way to check that this is not an environment issue.

**Comment by kinow** [comment]
Maybe we are allocating a whole process for the thread? I would expect Python to always have only one thread per process, due to the [GIL](https://wiki.python.org/moin/GlobalInterpreterLock). Unless we have some external library creating threads, if we wrap the threads in processes, or if we are using something different than cpython - https://docs.python.org/3/library/threading.html

**Comment by kinow** [comment]
In GitLab by @mgali on Apr 6, 2023, 16:36

Hi all,

Thanks a lot, sorry o couldn't give feedback until now. I no longer understand what you talk about but I see you are making progress.

Replying to @dbeltrankyl :

> I notice that:

> Each horizontal job starts at the same time.

> But between each job horizontal, the completed time is delayed is this normal? @mgali.

Does not look normal. All SIM jobs should last 3 minutes, with very small variations. So the variations you reported in COMPLETED time cannot be due to the jobs themselves.

**Comment by kinow** [comment]
In GitLab by @mgali on Apr 11, 2023, 08:57

Hi all,

I a launched exps a5v0 and a5v1 with above settings (MIN / MAX_WRAPPED = 5 / 25, without "_H") and with STRICT and MIXED policies, respectively. I confirm that MIXED works reasonably well whereas STRICT causes failure:

```
[CRITICAL] Wrapper couldn't be formed under strict POLICY due minimum limit not being reached: [wrappeable:1 < defined_min:1] [wrappeable_h:1 < defined_min_h:1]|[wrappeable_v:1 < defined_min_v:1]
Check your configuration: Only jobs_in_wrappers are active, check their dependencies. [eCode=7014]
```

I had to re-launch a5v0 with MIXED because it went inactive with this error after 4 generations (members). 

I'll post here a summary of the performance in this set of experiments as done before. Thanks!

**Comment by kinow** [comment]
In GitLab by @mgali on Apr 11, 2023, 12:42

On another note: experiments a5v0-2 were running fine. When I launched 2 more exps (a5v3,4) it seems all exps got stuck after a while. Will try to stop the last ones and eventually re-launch the first 3 which were close to finishing.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 13, 2023, 09:45

Thanks for the input, 

I've worked more on this, but I'm a bit stuck. The times as you said, are odd since if I launch the same templates standalone, I get these 3 mins. 

I think mpirun is still doing weird things even if I implicitly told him the number of resources, nodes, and machinefiles. I tried some other approaches in the wrapper, but it didn't appear to do anything. 


I'm now looking at env variables and adding debug flags to mpi ( in the copy I have locally ). If I find something relevant in the debugs that can't be solved within the wrapper, I'll ask to support because on the internet I found some bug reports that look similar to what is happening.

Maybe there is a bug in either slurm or mpi lib (pmoproxy) 

Could I get the slurm version somehow? sinfo -v seems restricted (@mcastril )

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 21, 2023, 12:53

I copy @gmontanepinto @agarci3 and @f-macchia because they usually employ horizontal or horitontal-vertical wrappers in their workflows

**Comment by kinow** [comment]
In GitLab by @mgali on May 12, 2023, 09:51

Hi all,

A litle update: during the last weeks I have been running experiments with just horizontal wrappers:

```
[wrapper]
TYPE = horizontal
JOBS_IN_WRAPPER = SIM
MACHINEFILES = STANDARD
MIN_WRAPPED = 5
MAX_WRAPPED = 25
POLICY = MIXED
```

They generally run fine, provided I do not launch several replicate experiments at the same time to avoid escalation of issues. Still, they run very slowly compared to 2 years ago. It seems this workflow with thousands of tiny jobs is penalized by the scheduler. I can clearly see that GA experiments run during the weekends and get stuck during working days.

Unless someone plans to work further on this workflow (@dbeltrankyl ? @mcastril ?), I will close the issue. I may reopen it in the future. Thanks for your efforts!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 12, 2023, 09:55

Hello @mgali ,

The issue is not easy to solve, and we also notice strange behavior in other workflows. We have open #1027 for this issue to keep track of it. 

I think we can close this one. If I find something new, I'll tag you in that issue
